### DISCUSSION TITLE
Enabling battery monitoring

### DISCUSSION THREAD

**Discussion by smedir** [jglathe/linux_ms_dev_kit#19] [Q&A] [answered]
Even the latest kernel, rc7 gives NaN to upowered, unable to monitor battery.

Any pointers?

Other things not working - sound, microphone, camera, bluetooth

**Comment by jglathe** [comment]
Hmm for the power/battery management to work (battery icon is white/green) there need to be a few prerequisites:

- User space must have pd-mapper and qrtr running. Both should be started by the image, works here, but I wasn't able to test this on x1e hardware yet.
- The remoteprocs (adsp at least) must be up and have their firmware loaded. For this the firmware needed to be copied  from Windows. In the same directory as the firmware there are some additional files, like some .elf (apparrently a device tree, interesting) and some .jsn files, they all need to be copied and available, too.
- The adsp and cdsp firmware (and accompanying files) all need to be in the initramfs, via /etc/initramfs-tools/hoolks/yoga7-firmware.

To check if the firmwares were actually loaded, you can use `sudo dmesg|grep Loaded` when you have my kernel parameters. My output on the X13s looks like this:
```
jglathe@x13s-jg:~$ sudo dmesg|grep Loaded
[    1.170571] remoteproc remoteproc1: Loaded FW: qcom/sc8280xp/LENOVO/21BX/qccdsp8280.mbn, sha256: 1b5fb275e5b33c6bd57aa77cc19cc338087ed8086a94053061578489e3c4dea1
[    1.193290] remoteproc remoteproc1: Loaded FW: qcom/sc8280xp/LENOVO/21BX/qccdsp8280.mbn, sha256: 1b5fb275e5b33c6bd57aa77cc19cc338087ed8086a94053061578489e3c4dea1
[    1.255186] remoteproc remoteproc0: Loaded FW: qcom/sc8280xp/LENOVO/21BX/qcadsp8280.mbn, sha256: 8d6bfb3e32cbfb8a775c9fc2a19e8a1044fc3c6abeb2302992cacf31c3971ef0
[    1.440828] remoteproc remoteproc0: Loaded FW: qcom/sc8280xp/LENOVO/21BX/qcadsp8280.mbn, sha256: 8d6bfb3e32cbfb8a775c9fc2a19e8a1044fc3c6abeb2302992cacf31c3971ef0
[    2.408062] msm_dpu ae01000.display-controller: Loaded FW: qcom/a660_sqe.fw, sha256: f911a4badf7b56719cb130492bc39f9edb85335f54e8b4983b5c67eb9e6204bb
[    2.408473] msm_dpu ae01000.display-controller: Loaded FW: qcom/a660_gmu.bin, sha256: 78f3c9650f38f5799019bd97e7c180d331239c1b321adc194422397403cf1b05
[    2.412706] msm_dpu ae01000.display-controller: Loaded FW: qcom/sc8280xp/LENOVO/21BX/qcdxkmsuc8280.mbn, sha256: 5e4c746dafa1f4892573bdc338bbf4c18665b8b32afb9255016f8543bd151094
[    4.912888] Loaded X.509 cert 'sforshee: 00b28ddf47aef9cea7'
[    4.913068] Loaded X.509 cert 'wens: 61c038651aabdcf94bd0ac7ff06c7248db18c600'
[    4.915003] platform regulatory.0: Loaded FW: regulatory.db, sha256: 31b14900eb290e982f52745deaec9b06dcaabb9293baf017c5118bfbef7d0a77
[    4.915239] platform regulatory.0: Loaded FW: regulatory.db.p7s, sha256: c142d2782f9ef6962888e92820d290287f0be179b3f388b1d72b541691d39fc2
[    5.190115] mhi mhi0: Loaded FW: ath11k/WCN6855/hw2.1/amss.bin, sha256: e12b23ddc4b8d2d2a10a651a5d6fdcd00f60fcae884d2cf5dad17627211fcdfd
[    5.395892] bluetooth hci0: Loaded FW: qca/hpbtfw21.tlv, sha256: a7494beaf00018942f695c85e5a7c7dc0aa017c8af2577007e343f4d80d038f4
[    5.953858] ath11k_pci 0006:01:00.0: Loaded FW: ath11k/WCN6855/hw2.1/board-2.bin, sha256: 146ad101f1894f8fdb0a9133394b4e294badceef72f6ce9d87cce2d9edc7d9a9
[    5.958947] ath11k_pci 0006:01:00.0: Loaded FW: ath11k/WCN6855/hw2.1/board-2.bin, sha256: 146ad101f1894f8fdb0a9133394b4e294badceef72f6ce9d87cce2d9edc7d9a9
[    5.969482] ath11k_pci 0006:01:00.0: Loaded FW: ath11k/WCN6855/hw2.1/board-2.bin, sha256: 146ad101f1894f8fdb0a9133394b4e294badceef72f6ce9d87cce2d9edc7d9a9
[    5.993591] ath11k_pci 0006:01:00.0: Loaded FW: ath11k/WCN6855/hw2.1/m3.bin, sha256: 0c590881870d0e6e98fc7d393ce05690e09287933b1b535e935bf5d98b77713f
[    6.023854] bluetooth hci0: Loaded FW: qca/hpnv21.bin, sha256: cccd47074913d1ed604ce4f56f9c1daf67ced38f88bff75fbf7863b3313864ce
[    9.345163] bluetooth hci0: Loaded FW: qca/hpbtfw21.tlv, sha256: a7494beaf00018942f695c85e5a7c7dc0aa017c8af2577007e343f4d80d038f4
[    9.960863] bluetooth hci0: Loaded FW: qca/hpnv21.bin, sha256: cccd47074913d1ed604ce4f56f9c1daf67ced38f88bff75fbf7863b3313864ce
[   10.260225] qcom-apm gprsvc:service:2:1: Loaded FW: qcom/sc8280xp/SC8280XP-LENOVO-X13S-tplg.bin, sha256: ee911c748deadc0191af50074b26ad177c8a2e78ff00e9578504b71d5d27b08a
[   42.905842] (NULL device *): Loaded FW: qcom/a660_sqe.fw, sha256: f911a4badf7b56719cb130492bc39f9edb85335f54e8b4983b5c67eb9e6204bb
[   42.905982] (NULL device *): Loaded FW: regulatory.db, sha256: 31b14900eb290e982f52745deaec9b06dcaabb9293baf017c5118bfbef7d0a77
[   42.906010] (NULL device *): Loaded FW: qca/hpnv21.bin, sha256: cccd47074913d1ed604ce4f56f9c1daf67ced38f88bff75fbf7863b3313864ce
[   42.906181] (NULL device *): Loaded FW: qcom/sc8280xp/SC8280XP-LENOVO-X13S-tplg.bin, sha256: ee911c748deadc0191af50074b26ad177c8a2e78ff00e9578504b71d5d27b08a
[   42.906283] (NULL device *): Loaded FW: qcom/a660_gmu.bin, sha256: 78f3c9650f38f5799019bd97e7c180d331239c1b321adc194422397403cf1b05
[   42.906402] (NULL device *): Loaded FW: qca/hpbtfw21.tlv, sha256: a7494beaf00018942f695c85e5a7c7dc0aa017c8af2577007e343f4d80d038f4
[   42.906508] (NULL device *): Loaded FW: regulatory.db.p7s, sha256: c142d2782f9ef6962888e92820d290287f0be179b3f388b1d72b541691d39fc2
[   42.908495] (NULL device *): Loaded FW: qcom/sc8280xp/LENOVO/21BX/qcdxkmsuc8280.mbn, sha256: 5e4c746dafa1f4892573bdc338bbf4c18665b8b32afb9255016f8543bd151094
[   42.908530] (NULL device *): Loaded FW: ath11k/WCN6855/hw2.1/m3.bin, sha256: 0c590881870d0e6e98fc7d393ce05690e09287933b1b535e935bf5d98b77713f
[   42.917160] (NULL device *): Loaded FW: qcom/sc8280xp/LENOVO/21BX/qccdsp8280.mbn, sha256: 1b5fb275e5b33c6bd57aa77cc19cc338087ed8086a94053061578489e3c4dea1
[   42.918042] (NULL device *): Loaded FW: ath11k/WCN6855/hw2.1/board-2.bin, sha256: 146ad101f1894f8fdb0a9133394b4e294badceef72f6ce9d87cce2d9edc7d9a9
[   42.926203] (NULL device *): Loaded FW: ath11k/WCN6855/hw2.1/amss.bin, sha256: e12b23ddc4b8d2d2a10a651a5d6fdcd00f60fcae884d2cf5dad17627211fcdfd
[   42.943264] (NULL device *): Loaded FW: qcom/sc8280xp/LENOVO/21BX/qcadsp8280.mbn, sha256: 8d6bfb3e32cbfb8a775c9fc2a19e8a1044fc3c6abeb2302992cacf31c3971ef0
[   43.772560] mhi mhi0: Loaded FW: ath11k/WCN6855/hw2.1/amss.bin, sha256: e12b23ddc4b8d2d2a10a651a5d6fdcd00f60fcae884d2cf5dad17627211fcdfd
[   44.827339] ath11k_pci 0006:01:00.0: Loaded FW: ath11k/WCN6855/hw2.1/board-2.bin, sha256: 146ad101f1894f8fdb0a9133394b4e294badceef72f6ce9d87cce2d9edc7d9a9
[   44.832194] ath11k_pci 0006:01:00.0: Loaded FW: ath11k/WCN6855/hw2.1/board-2.bin, sha256: 146ad101f1894f8fdb0a9133394b4e294badceef72f6ce9d87cce2d9edc7d9a9
[   44.842401] ath11k_pci 0006:01:00.0: Loaded FW: ath11k/WCN6855/hw2.1/board-2.bin, sha256: 146ad101f1894f8fdb0a9133394b4e294badceef72f6ce9d87cce2d9edc7d9a9
```
There are also ways to check if the remoteprocs are up and running. It involves sysfs / debugfs, will add when I find it.

Sound, mic, camera - no solution yet. BT - needs to be defined in the dts to be accessible. I will try this on the HP X14 with a known chipset (WCN6855). Getting the firmwares loaded (without killing boot via type-c) seemed to be a major hurdle in the past. I will document as I try this out.

**Comment by jglathe** [comment]
Update: I have come quite some distance on enabling the HP Omnibook X14. 

- all internal inputs work (kb, touchpad, touchscreen)
- WiFi and BT work
- Lid switch works
- firmwares get fetched from Windows
- battmgr (pd-mapper) works
- GPU acceleration works

Please have a look in the repos wdk2023_fw_fetcher and wdk2023_syshacks with branch hp-omnibook-x14 for clues.

**Reply by smedir** [reply]
Where exactly are the files, I cannot find them.

**Reply by jglathe** [reply]
Hmm [V7 of the image](https://drive.google.com/file/d/1Db_uowPtBtWgnfaoNlLBckjN7vzdKHMf/view?usp=sharing) is out, btw.

The fetch template file is [here](https://github.com/jglathe/wdk2023_fw_fetch/blob/jg/hp-omnibook-x14/bin/fetch_x1e80100_fw.sh). It works together with the [firmware hook in initramfs](https://github.com/jglathe/wdk2023_syshacks/blob/jg/hp-omnibook-x14/etc/initramfs-tools/hooks/x14-firmware). These are files intended for the HP Omnibook X14s, but can be easily adapted for the yoga slim 7x. The paths need to match those in the [dts of the yoga](https://github.com/anonymix007/linux-x1e/blob/wip/x1e80100-6.11/arch/arm64/boot/dts/qcom/x1e80100-lenovo-yoga-slim7x.dts).  

Do you have some model data for your Laptop? I might take a blind folded shot at Bluetooth.

**Reply by smedir** [reply]
This is very interesting. I am running a self compiled rc4 kernel. dmesg shows only the adreno.gpu firmware files are loaded. I do have some 8380.mbn firmware files in /lib/firmware in the initrd downloaded from the internet as well as the rootfs. They do not seem to be loaded. But touchscreen, keyboard, trackpad are working. Firmware initialized from prior windows boot? Running Debian Trixie.

I was unable to use your fetch template as fuse module was not compiled to mount ntfs-3g. However I followed the steps manually in Windows to extract all the *8380.mbn files and the *.efl and *.jsn files in the same folders. There are more than what are posted in your hook-

-adsp_dtbs.elf
adspr.jsn
adsps.jsn
adspua.jsn
battmgr.jsn
cdsp_dtbs.elf
cdspr.jsn
qcadsp8380.mbn
qcav1e8380.mbn
qccdsp8380.mbn
qcdxkmsuc8380.mbn
qcvk_icd_arm64x.json
qcvss8380.mbn
RCDS.bin

I added a *.json file and put in RCDS.bin just for good measure.

I used your initramfs hooks  template with the corrected file paths to make a new initrd.img image.

Booting with the image, the kernel boots successfully, Userland and systemd starts loading services. At some service, it scrolls too fast- the kernel starts giving tons of errors- I manage to glance at some of the messages.

It seems some service is loading qcadsp8380.mbn, at which point the kernel starts panicking and the system slows to a crawl, still trying to load other services.. but you can wait for a long time, and nothing happens.

**Reply by smedir** [reply]
Looking at the dts file, it seems only the following files are loaded-

qcdxkmsuc8380.mbn
qcadsp8380.mbn
qccdsp8380.mbn
cdsp_dtbs.elf
adsp_dtbs.elf

**Reply by jglathe** [reply]
the *.jsn files are required for userland pd-mapper at least - I guess even for in-kernel pd-mapper they are needed. They are loaded by adsp (without showing much of a log).
V7 of the image is 6.11, can boot the yoga 7 slim and has ntfs-3g compiled, maybe just try with that? The order you see in my firmware hook is what I use on the HP Omnibook X14.

**Reply by smedir** [reply, marked as answer]
The json files in intird.img did the trick.

Had to install pd-mapper binary, as well as activate the service with a new service file, and the battery level and charging shows.

Here is the complete hook file I used in intramfs to generate the correct initrd.img-

````
set -e

PREREQ=""

prereqs()
{
	echo "$PREREQ"
}

case \\$1 in
# get pre-requisites
prereqs)
	prereqs
	exit 0
	;;
esac

. /usr/share/initramfs-tools/hook-functions

#mv /lib/firmware/qcom/x1e80100/LENOVO/83ED/qcadsp8380.mbn.disabled /lib/firmware/qcom/x1e80100/LENOVO/83ED/qcadsp8380.mbn

# Define a list of firmware files to be included

FIRMWARE_FILES="\
qcom/x1e80100/LENOVO/83ED/adsp_dtbs.elf \
qcom/x1e80100/LENOVO/83ED/adspr.jsn \
qcom/x1e80100/LENOVO/83ED/adsps.jsn \
qcom/x1e80100/LENOVO/83ED/adspua.jsn \
qcom/x1e80100/LENOVO/83ED/battmgr.jsn \
qcom/x1e80100/LENOVO/83ED/cdspr.jsn \
qcom/x1e80100/LENOVO/83ED/cdsp_dtbs.elf \
qcom/x1e80100/LENOVO/83ED/qccdsp8380.mbn \
qcom/x1e80100/LENOVO/83ED/qcadsp8380.mbn \
qcom/x1e80100/LENOVO/83ED/qcdxkmsuc8380.mbn \
qcom/x1e80100/LENOVO/83ED/qcav1e8380.mbn \
qcom/x1e80100/LENOVO/83ED/qcvss8380.mbn \
qcom/x1e80100/LENOVO/83ED/qcvk_icd_arm64x.json \
qcom/gen70500_gmu.bin \
qcom/gen70500_sqe.fw"

# Copy each firmware file to initramfs
for file in $FIRMWARE_FILES; do
    dir=$(dirname "$file")
    mkdir -p "${DESTDIR}/lib/firmware/${dir}"
    cp "/lib/firmware/${file}" "${DESTDIR}/lib/firmware/${dir}/"
done

#mv /lib/firmware/qcom/x1e80100/LENOVO/83ED/qcadsp8380.mbn /lib/firmware/qcom/x1e80100/LENOVO/83ED/qcadsp8380.mbn.disabled

````

Summary:

Make sure all firmware files and associated .json files are present in the initrd.img with hook files provided by ghlathe
Compile pd-mapper userland binary if not provided by your distribution.
Start pd-mapper as a service with the included systemd file
Make sure kernel pdmapper is present- `lsmod | grep qcom_pd_mapper`

Battery manger shows up correctly including charging discharging status!
